solr/core/src/java/org/apache/solr/cloud/api/collections/OverseerCollectionMessageHandler.java

```diff
@@ -672,35 +636,35 @@ void cleanupCollection(String collectionName, NamedList results) throws Exceptio
     commandMap.get(DELETE).call(zkStateReader.getClusterState(), new ZkNodeProps(props), results);
   }
 
-  Map<String, Replica> waitToSeeReplicasInState(String collectionName, Collection<String> coreNames) throws InterruptedException {
-    assert coreNames.size() > 0;
-    Map<String, Replica> result = new HashMap<>();
-    TimeOut timeout = new TimeOut(Integer.getInteger("solr.waitToSeeReplicasInStateTimeoutSeconds", 120), TimeUnit.SECONDS, timeSource); // could be a big cluster
-    while (true) {
-      DocCollection coll = zkStateReader.getClusterState().getCollection(collectionName);
-      for (String coreName : coreNames) {
-        if (result.containsKey(coreName)) continue;
-        for (Slice slice : coll.getSlices()) {
-          for (Replica replica : slice.getReplicas()) {
-            if (coreName.equals(replica.getStr(ZkStateReader.CORE_NAME_PROP))) {
-              result.put(coreName, replica);
-              break;
+  Map<String, Replica> waitToSeeReplicasInState(String collectionName, Collection<String> coreNames) {
+    final Map<String, Replica> result = new HashMap<>();
+    int timeout = Integer.getInteger("solr.waitToSeeReplicasInStateTimeoutSeconds", 120); // could be a big cluster
+    try {
+      zkStateReader.waitForState(collectionName, timeout, TimeUnit.SECONDS, c -> {
+        // todo this is ugly, but I'm not sure there is a better way to fix it?
```

Comment:
Can't we iterate shards/replicas, and for each one check if coreNames .contains(replica.getStr(ZkStateReader.CORE_NAME_PROP)). Maybe make a set with all the elements in coreNames and remove them as you find them, and break if empty? I guess it depends on how big coreNames will be

Reply:
I think I found something that makes sense.